File: algorithm/typescript/BOJ2667.ts
// problem : https://www.acmicpc.net/problem/2667

import fs = require("fs");

const input: string[] = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
const N: number = Number(input[0]);
const map: number[][] = new Array();
const visited: boolean[][] = new Array();
const answer: number[] = new Array();
const dx: number[] = [-1, 1, 0, 0];
const dy: number[] = [0, 0, -1, 1];
let count: number = 1;

for (let i = 1; i <= N; i++) {
  const mapRow: number[] = input[i].split("").map(Number);
  const visitedRow: boolean[] = new Array(N).fill(false);
  map.push(mapRow);
  visited.push(visitedRow);
}

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    if (map[i][j] === 1 && !visited[i][j]) {
      count = 1;
      dfs(i, j);
      answer.push(count);
    }
  }
}

function dfs(x: number, y: number): void {
  visited[x][y] = true;

  for (let i = 0; i < 4; i++) {
    let nx: number = x + dx[i];
    let ny: number = y + dy[i];

    if (nx < 0 || ny < 0 || nx >= N || ny >= N || map[nx][ny] === 0 || visited[nx][ny]) {
      continue;
    }

    count++;
    dfs(nx, ny);
  }
}

console.log(answer.length);
answer.sort((a, b) => {
  return a - b;
});
console.log(answer.join("\n"));

File: algorithm/python/BOJ2667.py
# problem : https://www.acmicpc.net/problem/2667

import sys

N = int(sys.stdin.readline().rstrip())
grid = list()
answer = list()
visited = list()
count = 1
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

for i in range(0, N):
    grid_row = list(map(int, sys.stdin.readline().rstrip()))
    visit_row = [False] * N
    grid.append(grid_row)
    visited.append(visit_row)


def dfs(x: int, y: int):
    global count
    visited[x][y] = True

    for i in range(0, 4):
        new_x = x + dx[i]
        new_y = y + dy[i]

        if new_x < 0 or new_x >= N or new_y < 0 or new_y >= N or visited[new_x][new_y] or grid[new_x][new_y] == 0:
            continue

        count += 1
        dfs(new_x, new_y)


for i in range(0, N):
    for j in range(0, N):
        if grid[i][j] == 1 and not visited[i][j]:
            count = 1
            dfs(i, j)
            answer.append(count)

print(len(answer))
answer.sort()
for item in answer:
    print(item)
